core/src/main/java/org/jruby/runtime/TraceEventManager.java

```diff
@@ -0,0 +1,273 @@
+                if (eventHook.isInterestedInEvent(event)) {
+                    IRubyObject klass = context.nil;
+                    if (type instanceof RubyModule) {
+                        if (((RubyModule) type).isIncluded()) {
+                            klass = ((RubyModule) type).getOrigin();
+                        } else if (((RubyModule) type).isSingleton()) {
+                            klass = ((MetaClass) type).getAttached();
+                        }
+                    }
+                    eventHook.event(context, event, file, line, name, klass);
+                }
+            }
+        }
+    }
+
+    public MutableCallSite getCallReturnSite() {
+        return callTrace;
+    }
+
+    public MutableCallSite getBCallBReturnSite() {
+        return bcallTrace;
+    }
+
+    public boolean hasEventHooks() {
+        return hasEventHooks;
+    }
+
+    public static class CallTraceFuncHook extends EventHook {
+        private RubyProc traceFunc;
+        private final ThreadContext thread; // if non-null only call traceFunc if it is from this thread.
+
+        public CallTraceFuncHook(ThreadContext context) {
+            this.thread = context;
+        }
+
+        public void setTraceFunc(RubyProc traceFunc) {
+            this.traceFunc = traceFunc;
+        }
+
+        public void eventHandler(ThreadContext context, String eventName, String file, int line, String name, IRubyObject type) {
+            if (context.isWithinTrace()) return;
+            if (thread != null && thread != context) return;
+
+            if (file == null) file = "(ruby)";
+            if (type == null) type = context.nil;
+
+            Ruby runtime = context.runtime;
+            RubyBinding binding = RubyBinding.newBinding(runtime, context.currentBinding());
+
+            // FIXME: Ultimately we should be getting proper string for this event type
+            switch (eventName) {
+                case "c_return":
+                    eventName = "c-return";
+                    break;
+                case "c_call":
+                    eventName = "c-call";
+                    break;
+            }
+            ;
+
+            context.preTrace();
+            try {
+                traceFunc.call(context, new IRubyObject[]{
+                        runtime.newString(eventName), // event name
+                        runtime.newString(file), // filename
+                        runtime.newFixnum(line), // line numbers should be 1-based
+                        name != null ? runtime.newSymbol(name) : runtime.getNil(),
+                        binding,
+                        type
+                });
+            } finally {
+                context.postTrace();
+            }
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (!(other instanceof CallTraceFuncHook)) return false;
+
+            return super.equals(other) && thread == ((CallTraceFuncHook) other).thread;
```

Comment:
this existed before and I recall it being confusing, 
the intent of checking `this == other` and than doing more checks.

maybe 
```suggestion
            return traceFunc == ((CallTraceFuncHook) other).traceFunc && thread == ((CallTraceFuncHook) other).thread;
```